### DISCUSSION TITLE
EntityAdapter | removeMany with Predicate Function

### DISCUSSION THREAD

**Discussion by xbaun** [reduxjs/redux-toolkit#2695] [Q&A] [answered]
Hi, 

I search for a way to dispatch a `removeMany` action with a predicate function to filter for entities that should be removed. Is this scenario somehow covered by the EntityAdapter?

Thanks

**Comment by markerikson** [comment, marked as answer]
Not directly, no. 

The primary issue here is that functions are not serializable, and thus they don't belong in an action:

- https://redux.js.org/style-guide/#do-not-put-non-serializable-values-in-state-or-actions

`@ngrx/entity` _does_ actually have a `removeMany(predicate)` option, and RTK's `createEntityAdapter` started as a copy of that package.  But, when we reworked the methods to be usable as reducers directly, we had to remove that method to keep usage consistent with Redux patterns.

The best option would be to do a `.ids.filter()`, look up the values from the `.entities` field, come up with an array of IDs to remove, and pass those to `adapter.removeMany(state, ids)`.

**Reply by xbaun** [reply]
Thank you for the answer. 

My implementation was exactly as you suggested. What is rather messy about this is that a separate action has to be written for each filter on top of the Entity Adapter. I was hoping there would be a nicer way to do this.
